llmfoundry/callbacks/curriculum_learning_callback.py

```diff
@@ -34,20 +40,188 @@ class CurriculumLearning(CallbackWithConfig):
         dataset_index (int): The index of the dataset currently being used.
     """
 
-    def __init__(self, train_config: Dict, dataset_index: int):
-        self.dataset_index = dataset_index
-        self.saved_dataset_index = 0
-        self.all_dataset_configs = []
-        self.current_dataset_state = {}
-        # The current dataset config is resolved and passed in train.py
-        self.current_dataset_config = train_config['train_loader']
+    def __init__(
+        self,
+        train_config: dict[str, Any],
+        schedule: list[dict[str, Any]],
+    ):
+        from llmfoundry.utils.builders import build_tokenizer
+        from llmfoundry.utils.config_utils import calculate_batch_size_info
+
+        # Ensure all duration values are positive
+        # Ensure all duration units are in epochs or tokens
+        self._schedule = schedule
+        if len(self._schedule) == 0:
+            raise ValueError('The schedule must have at least one datamix.')
+        for index, datamix in enumerate(self._schedule):
+            self._validate_data_mix(datamix)
+
+            if (
+                index == 0 and
+                train_config['train_loader'] != datamix['train_loader']
+            ):
+                raise ValueError((
+                    'The first datamix in the schedule must match the '
+                    'train_loader in the train_config.'
+                ))
+
+        self._schedule_index = 0
+
+        # Copied from llmfoundry/utils/config_utils.py
+        self.device_train_batch_size, _, _ = calculate_batch_size_info(
+            train_config['global_train_batch_size'],
+            train_config['device_train_microbatch_size'],
+            data_replication_degree=1,
+        )
+
+        # Copied from scripts/train/train.py
+        tokenizer_name = train_config['tokenizer']['name']
+        tokenizer_kwargs = train_config['tokenizer'].get('kwargs', {})
+        self.tokenizer = build_tokenizer(tokenizer_name, tokenizer_kwargs)
```

Comment:
Why is this needed on init? Can we not pull this on `Event.INIT`?

Reply:
You can do this during Event.INIT but then you'd need to save the train_config in the callback during init. It seems easier to do this during init

Reply:
Why is it easier during __init__?

Reply:
You get passed the train_config during init. If you want to do it during Event.init then you need to save the train_config as a class attribute in init then access it during Event.init to create more class attributes. It seems better to just create it in init and not have to save train_config.